[Issue]: Offline connection does not support method, toString

### Description

I am using the Docker variant of SchemaCrawler 16.15.11 with a MySQL 5.7 database and tried the "offline schema" feature. Generating the offline schema as JSON or YAML works very well. But generating a diagram of that offline schema leads to the following exception:

```
Nov 08, 2021 8:36:19 AM schemacrawler.tools.commandline.utility.CommandLineLoggingUtility logFatalStackTrace
SEVERE: null
java.lang.reflect.UndeclaredThrowableException
        at jdk.proxy2/jdk.proxy2.$Proxy13.toString(Unknown Source)
        at java.base/java.util.Formatter$FormatSpecifier.printString(Formatter.java:3057)
        at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:2934)
        at java.base/java.util.Formatter.format(Formatter.java:2690)
        at java.base/java.util.Formatter.format(Formatter.java:2625)
        at us.fatehi.utility.string.StringFormat.get(StringFormat.java:52)
        at us.fatehi.utility.string.StringFormat.get(StringFormat.java:35)
        at java.logging/java.util.logging.Logger.log(Logger.java:1050)
        at schemacrawler.tools.databaseconnector.DatabaseConnectionSource.getConnection(DatabaseConnectionSource.java:185)
        at schemacrawler.tools.databaseconnector.DatabaseConnectionSource.get(DatabaseConnectionSource.java:84)
        at schemacrawler.tools.databaseconnector.DatabaseConnectionSource.get(DatabaseConnectionSource.java:53)
        at schemacrawler.tools.commandline.command.ConnectCommand.loadSchemaRetrievalOptionsBuilder(ConnectCommand.java:186)
        at schemacrawler.tools.commandline.command.ConnectCommand.run(ConnectCommand.java:126)
        at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.executeCommandLine(SchemaCrawlerCommandLine.java:99)
        at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.execute(SchemaCrawlerCommandLine.java:67)
        at schemacrawler.Main.main(Main.java:73)
Caused by: java.sql.SQLFeatureNotSupportedException: Offline connection does not support method, toString
        at schemacrawler.tools.offline.jdbc.OfflineConnectionUtility$OfflineConnectionInvocationHandler.invoke(OfflineConnectionUtility.java:66)
        ... 16 more` 
```


### Reproduce

These are my command line options for generating the offline schema:
`--info-level=maximum --command serialize --portable-names=true --sort-columns=false --output-format=yaml --output-file=er_graph.yml --load-row-counts=false --log-level=ALL`

These are my command line options for using the offline schema to generate a diagram:
`--server=offline --database=/home/schcrwlr/er_graph.yml --info-level=standard --command=details --portable-names=true --sort-columns=false --output-format=png --output-file=er_graph.png --load-row-counts=false --log-level=ALL`

### Relevant log output

see description


### SchemaCrawler Version

16.15.11

### Java Version

Oracle Corporation OpenJDK 64-Bit Server VM 17-ea+14

### Operating System and Version

Windows 10

### Relational Database System and Version

MySQL 5.7 (AWS Aurora)

### JDBC Driver and Version

Unknown, running inside docker

@sualeh Thank you for your hint. But event with --output-format=ser the error stays absolutely the same:

`Nov 09, 2021 6:59:40 AM schemacrawler.tools.commandline.utility.CommandLineLoggingUtility logSafeArguments
INFO: Command line:
--server=offline
--database=/home/schcrwlr/er_graph.ser
--info-level=standard
--command=details
--portable-names=true
--sort-columns=false
--output-format=png
--output-file=er_graph.png
--load-row-counts=false
--log-level=ALL
Nov 09, 2021 6:59:40 AM schemacrawler.tools.commandline.utility.CommandLineLoggingUtility logFatalStackTrace
SEVERE: null
java.lang.reflect.UndeclaredThrowableException
        at jdk.proxy2/jdk.proxy2.$Proxy13.toString(Unknown Source)
        at java.base/java.util.Formatter$FormatSpecifier.printString(Formatter.java:3057)
        at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:2934)
        at java.base/java.util.Formatter.format(Formatter.java:2690)
        at java.base/java.util.Formatter.format(Formatter.java:2625)
        at us.fatehi.utility.string.StringFormat.get(StringFormat.java:52)
        at us.fatehi.utility.string.StringFormat.get(StringFormat.java:35)
        at java.logging/java.util.logging.Logger.log(Logger.java:1050)
        at schemacrawler.tools.databaseconnector.DatabaseConnectionSource.getConnection(DatabaseConnectionSource.java:185)
        at schemacrawler.tools.databaseconnector.DatabaseConnectionSource.get(DatabaseConnectionSource.java:84)
        at schemacrawler.tools.databaseconnector.DatabaseConnectionSource.get(DatabaseConnectionSource.java:53)
        at schemacrawler.tools.commandline.command.ConnectCommand.loadSchemaRetrievalOptionsBuilder(ConnectCommand.java:186)
        at schemacrawler.tools.commandline.command.ConnectCommand.run(ConnectCommand.java:126)
        at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.executeCommandLine(SchemaCrawlerCommandLine.java:99)
        at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.execute(SchemaCrawlerCommandLine.java:67)
        at schemacrawler.Main.main(Main.java:73)
Caused by: java.sql.SQLFeatureNotSupportedException: Offline connection does not support method, toString
        at schemacrawler.tools.offline.jdbc.OfflineConnectionUtility$OfflineConnectionInvocationHandler.invoke(OfflineConnectionUtility.java:66)
        ... 16 more

SchemaCrawler 16.15.11`

--- a/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/OfflineCatalogLoader.java
+++ b/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/OfflineCatalogLoader.java
@@ -35,7 +35,7 @@ public final class OfflineCatalogLoader extends BaseCatalogLoader {
   private static final Logger LOGGER = Logger.getLogger(OfflineCatalogLoader.class.getName());
 
   public OfflineCatalogLoader() {
-    super(new CommandDescription("offlineloader", "Loader for offline databases"), -1);
+    super(new CommandDescription("offlineloader", "Loader for offline catalog snapshots"), -1);
   }
 
   @Override
@@ -67,7 +67,8 @@ public final class OfflineCatalogLoader extends BaseCatalogLoader {
       final Path offlineDatabasePath = dbConnection.getOfflineDatabasePath();
       LOGGER.log(
           Level.CONFIG,
-          new StringFormat("Reading serialized database file <%s>", offlineDatabasePath));
+          new StringFormat(
+              "Reading serialized offline catalog snapshot <%s>", offlineDatabasePath));
 
       try (final InputStream inputFileStream = Files.newInputStream(offlineDatabasePath)) {
         final JavaSerializedCatalog deserializedCatalog =
@@ -76,7 +77,7 @@ public final class OfflineCatalogLoader extends BaseCatalogLoader {
       }
       reduceCatalog(catalog);
     } catch (final IOException | SQLException e) {
-      throw new SchemaCrawlerException("Could not load offline database", e);
+      throw new SchemaCrawlerException("Could not load offline catalog snapshot", e);
     }
 
     setCatalog(catalog);

--- a/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/OfflineCatalogLoader.java
+++ b/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/OfflineCatalogLoader.java
@@ -6,11 +6,14 @@ import static schemacrawler.filter.ReducerFactory.getSequenceReducer;
 import static schemacrawler.filter.ReducerFactory.getSynonymReducer;
 import static schemacrawler.filter.ReducerFactory.getTableReducer;
 
-import java.io.FileInputStream;
 import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.sql.Connection;
 import java.sql.SQLException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 import schemacrawler.schema.Catalog;
 import schemacrawler.schema.Reducible;
@@ -25,8 +28,11 @@ import schemacrawler.tools.catalogloader.BaseCatalogLoader;
 import schemacrawler.tools.executable.CommandDescription;
 import schemacrawler.tools.formatter.serialize.JavaSerializedCatalog;
 import schemacrawler.tools.offline.jdbc.OfflineConnection;
+import us.fatehi.utility.string.StringFormat;
 
 public final class OfflineCatalogLoader extends BaseCatalogLoader {
+
+  private static final Logger LOGGER = Logger.getLogger(OfflineCatalogLoader.class.getName());
 
   public OfflineCatalogLoader() {
     super(new CommandDescription("offlineloader", "Loader for offline databases"), -1);
@@ -59,8 +65,11 @@ public final class OfflineCatalogLoader extends BaseCatalogLoader {
       }
 
       final Path offlineDatabasePath = dbConnection.getOfflineDatabasePath();
-      try (final FileInputStream inputFileStream =
-          new FileInputStream(offlineDatabasePath.toFile())) {
+      LOGGER.log(
+          Level.CONFIG,
+          new StringFormat("Reading serialized database file <%s>", offlineDatabasePath));
+
+      try (final InputStream inputFileStream = Files.newInputStream(offlineDatabasePath)) {
         final JavaSerializedCatalog deserializedCatalog =
             new JavaSerializedCatalog(inputFileStream);
         catalog = deserializedCatalog.getCatalog();

--- a/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/OfflineDatabaseConnector.java
+++ b/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/OfflineDatabaseConnector.java
@@ -56,7 +56,7 @@ public final class OfflineDatabaseConnector extends DatabaseConnector {
         .addOption(
             "server",
             String.class,
-            "--server=offline%n" + "Loads SchemaCrawler plug-in for offline snapshots")
+            "--server=offline%n" + "Loads SchemaCrawler plug-in for offline catalog snapshots")
         .addOption("host", String.class, "Should be omitted")
         .addOption("port", Integer.class, "Should be omitted")
         .addOption(

--- a/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/jdbc/OfflineConnectionUtility.java
+++ b/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/jdbc/OfflineConnectionUtility.java
@@ -35,6 +35,7 @@ import java.lang.reflect.InvocationHandler;
 import java.lang.reflect.Method;
 import java.nio.file.Path;
 import java.sql.SQLFeatureNotSupportedException;
+import java.util.Collections;
 
 public class OfflineConnectionUtility {
 
@@ -62,17 +63,23 @@ public class OfflineConnectionUtility {
           return false;
         case "isValid":
           return true;
+        case "getTypeMap":
+          return Collections.emptyMap();
         case "toString":
-          return String.format("SchemaCrawler Offline Database Connection");
+          return String.format(
+              "schemacrawler.tools.offline.jdbc.OfflineConnection@%s",
+              offlineDatabasePath.hashCode());
         default:
           throw new SQLFeatureNotSupportedException(
-              "Offline connection does not support method, " + methodName, "HYC00");
+              String.format(
+                  "Offline catalog snapshot connection does not support method <%s>", methodName),
+              "HYC00");
       }
     }
   }
 
   public static OfflineConnection newOfflineConnection(final Path offlineDatabasePath) {
-    requireNonNull(offlineDatabasePath, "No offline database path provided");
+    requireNonNull(offlineDatabasePath, "No offline catalog snapshot path provided");
 
     final Path absoluteOfflineDatabasePath = offlineDatabasePath.toAbsolutePath();
     return (OfflineConnection)

--- a/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/jdbc/OfflineConnectionUtility.java
+++ b/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/jdbc/OfflineConnectionUtility.java
@@ -62,6 +62,8 @@ public class OfflineConnectionUtility {
           return false;
         case "isValid":
           return true;
+        case "toString":
+          return String.format("SchemaCrawler Offline Database Connection");
         default:
           throw new SQLFeatureNotSupportedException(
               "Offline connection does not support method, " + methodName, "HYC00");
@@ -72,11 +74,12 @@ public class OfflineConnectionUtility {
   public static OfflineConnection newOfflineConnection(final Path offlineDatabasePath) {
     requireNonNull(offlineDatabasePath, "No offline database path provided");
 
+    final Path absoluteOfflineDatabasePath = offlineDatabasePath.toAbsolutePath();
     return (OfflineConnection)
         newProxyInstance(
             OfflineConnectionUtility.class.getClassLoader(),
             new Class[] {OfflineConnection.class},
-            new OfflineConnectionInvocationHandler(offlineDatabasePath));
+            new OfflineConnectionInvocationHandler(absoluteOfflineDatabasePath));
   }
 
   private OfflineConnectionUtility() {

--- a/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/jdbc/OfflineDriver.java
+++ b/schemacrawler-offline/src/main/java/schemacrawler/tools/offline/jdbc/OfflineDriver.java
@@ -53,7 +53,7 @@ public class OfflineDriver implements Driver {
     try {
       DriverManager.registerDriver(new OfflineDriver());
     } catch (final SQLException e) {
-      LOGGER.log(Level.SEVERE, "Cannot register offline driver", e);
+      LOGGER.log(Level.SEVERE, "Cannot register SchemaCrawler Offline Catalog Snapshot driver", e);
     }
   }
 

--- a/schemacrawler-offline/src/test/resources/command_line_help_output/offline.help.txt
+++ b/schemacrawler-offline/src/test/resources/command_line_help_output/offline.help.txt
@@ -6,6 +6,7 @@
       --host=<host>       Should be omitted
       --port=<port>       Should be omitted
       --server=<server>   --server=offline
-                          Loads SchemaCrawler plug-in for offline snapshots
+                          Loads SchemaCrawler plug-in for offline catalog
+                            snapshots
 
 

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/catalogloader/CatalogLoaderRegistry.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/catalogloader/CatalogLoaderRegistry.java
@@ -50,39 +50,27 @@ public final class CatalogLoaderRegistry {
 
   public Collection<PluginCommand> getCommandLineCommands() {
     final Collection<PluginCommand> commandLineCommands = new HashSet<>();
-    try {
-      for (final CatalogLoader catalogLoader : instantiateCatalogLoaders()) {
-        commandLineCommands.add(catalogLoader.getCommandLineCommand());
-      }
-    } catch (final SchemaCrawlerException e) {
-      throw new SchemaCrawlerRuntimeException("Could not load catalog loaders", e);
+    for (final CatalogLoader catalogLoader : instantiateCatalogLoaders()) {
+      commandLineCommands.add(catalogLoader.getCommandLineCommand());
     }
     return commandLineCommands;
   }
 
   public Collection<PluginCommand> getHelpCommands() {
     final Collection<PluginCommand> commandLineCommands = new HashSet<>();
-    try {
-      for (final CatalogLoader catalogLoader : instantiateCatalogLoaders()) {
-        commandLineCommands.add(catalogLoader.getHelpCommand());
-      }
-    } catch (final SchemaCrawlerException e) {
-      throw new SchemaCrawlerRuntimeException("Could not load catalog loaders", e);
+    for (final CatalogLoader catalogLoader : instantiateCatalogLoaders()) {
+      commandLineCommands.add(catalogLoader.getHelpCommand());
     }
     return commandLineCommands;
   }
 
   public Collection<CommandDescription> getSupportedCatalogLoaders() {
     final Collection<CommandDescription> commandLineCommands = new HashSet<>();
-    try {
-      for (final CatalogLoader catalogLoader : instantiateCatalogLoaders()) {
-        final CommandDescription commandDescription = catalogLoader.getCommandDescription();
-        commandLineCommands.add(
-            new CommandDescription(
-                commandDescription.getName(), commandDescription.getDescription()));
-      }
-    } catch (final SchemaCrawlerException e) {
-      throw new SchemaCrawlerRuntimeException("Could not load catalog loaders", e);
+    for (final CatalogLoader catalogLoader : instantiateCatalogLoaders()) {
+      final CommandDescription commandDescription = catalogLoader.getCommandDescription();
+      commandLineCommands.add(
+          new CommandDescription(
+              commandDescription.getName(), commandDescription.getDescription()));
     }
     return commandLineCommands;
   }
@@ -92,7 +80,7 @@ public final class CatalogLoaderRegistry {
     return new ChainedCatalogLoader(chainedCatalogLoaders);
   }
 
-  private List<CatalogLoader> instantiateCatalogLoaders() throws SchemaCrawlerException {
+  private List<CatalogLoader> instantiateCatalogLoaders() {
 
     final List<CatalogLoader> catalogLoaderRegistry = new ArrayList<>();
 
@@ -106,7 +94,7 @@ public final class CatalogLoaderRegistry {
         catalogLoaderRegistry.add(catalogLoader);
       }
     } catch (final Throwable e) {
-      throw new SchemaCrawlerException("Could not load catalog loader registry", e);
+      throw new SchemaCrawlerRuntimeException("Could not load catalog loader registry", e);
     }
 
     Collections.sort(catalogLoaderRegistry);

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/DatabaseConnectorRegistry.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/DatabaseConnectorRegistry.java
@@ -43,8 +43,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.ServiceLoader;
 import java.util.logging.Level;
-
 import java.util.logging.Logger;
+
 import schemacrawler.schemacrawler.DatabaseServerType;
 import schemacrawler.schemacrawler.SchemaCrawlerRuntimeException;
 import schemacrawler.tools.executable.commandline.PluginCommand;
@@ -57,8 +57,7 @@ import us.fatehi.utility.string.StringFormat;
  */
 public final class DatabaseConnectorRegistry implements Iterable<DatabaseServerType> {
 
-  private static final Logger LOGGER =
-      Logger.getLogger(DatabaseConnectorRegistry.class.getName());
+  private static final Logger LOGGER = Logger.getLogger(DatabaseConnectorRegistry.class.getName());
 
   private static DatabaseConnectorRegistry databaseConnectorRegistrySingleton;
 
@@ -92,8 +91,8 @@ public final class DatabaseConnectorRegistry implements Iterable<DatabaseServerT
               Level.CONFIG,
               e,
               new StringFormat(
-            "Could not load database connector, %s=%s",
-            databaseSystemIdentifier, databaseConnector.getClass().getName()));
+                  "Could not load database connector, %s=%s",
+                  databaseSystemIdentifier, databaseConnector.getClass().getName()));
         }
       }
     } catch (final Exception e) {
@@ -127,7 +126,7 @@ public final class DatabaseConnectorRegistry implements Iterable<DatabaseServerT
           buffer.append(System.lineSeparator());
         }
       }
-    } catch (final Exception e) {
+    } catch (final Throwable e) {
       throw new SchemaCrawlerRuntimeException("Could not load database drivers", e);
     }
     if (log) {

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/UnknownDatabaseConnector.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/UnknownDatabaseConnector.java
@@ -31,6 +31,7 @@ import java.util.regex.Pattern;
 
 import schemacrawler.schemacrawler.DatabaseServerType;
 import schemacrawler.schemacrawler.SchemaCrawlerException;
+import schemacrawler.schemacrawler.SchemaCrawlerRuntimeException;
 import us.fatehi.utility.PropertiesUtility;
 
 final class UnknownDatabaseConnector extends DatabaseConnector {
@@ -62,21 +63,25 @@ final class UnknownDatabaseConnector extends DatabaseConnector {
     final DatabaseConnectionSource databaseConnectionSource =
         super.newDatabaseConnectionSource(connectionOptions);
 
+    final String url = databaseConnectionSource.getConnectionUrl();
+    failIfSchemaCrawlerPluginsNotAvailable(url);
+
+    return databaseConnectionSource;
+  }
+
+  private void failIfSchemaCrawlerPluginsNotAvailable(final String url) {
     final String withoutDatabasePlugin =
         PropertiesUtility.getSystemConfigurationProperty(
             "SC_IGNORE_MISSING_DATABASE_PLUGIN", Boolean.FALSE.toString());
     if (!Boolean.valueOf(withoutDatabasePlugin)) {
       // Check if SchemaCrawler database plugin is in use
-      final String url = databaseConnectionSource.getConnectionUrl();
       for (final Pattern pattern : patterns) {
         if (pattern.matcher(url).matches()) {
-          throw new SchemaCrawlerException(
+          throw new SchemaCrawlerRuntimeException(
               String.format(
                   "SchemaCrawler database plugin should be on the CLASSPATH for <%s>", url));
         }
       }
     }
-
-    return databaseConnectionSource;
   }
 }

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/executable/CommandRegistry.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/executable/CommandRegistry.java
@@ -56,17 +56,13 @@ public final class CommandRegistry {
   private static CommandRegistry commandRegistrySingleton;
 
   public static CommandRegistry getCommandRegistry() {
-    try {
-      if (commandRegistrySingleton == null) {
-        commandRegistrySingleton = new CommandRegistry();
-      }
-    } catch (final SchemaCrawlerException e) {
-      throw new SchemaCrawlerRuntimeException("Cannot load SchemaCrawler commands", e);
+    if (commandRegistrySingleton == null) {
+      commandRegistrySingleton = new CommandRegistry();
     }
     return commandRegistrySingleton;
   }
 
-  private static List<CommandProvider> loadCommandRegistry() throws SchemaCrawlerException {
+  private static List<CommandProvider> loadCommandRegistry() {
 
     final List<CommandProvider> commandProviders = new ArrayList<>();
 
@@ -81,8 +77,8 @@ public final class CommandRegistry {
                 commandProvider.getSupportedCommands(), commandProvider.getClass().getName()));
         commandProviders.add(commandProvider);
       }
-    } catch (final Exception e) {
-      throw new SchemaCrawlerException("Could not load extended command registry", e);
+    } catch (final Throwable e) {
+      throw new SchemaCrawlerRuntimeException("Could not load extended command registry", e);
     }
 
     return commandProviders;
@@ -90,7 +86,7 @@ public final class CommandRegistry {
 
   private final List<CommandProvider> commandRegistry;
 
-  private CommandRegistry() throws SchemaCrawlerException {
+  private CommandRegistry() {
     commandRegistry = loadCommandRegistry();
   }
 

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/executable/SchemaCrawlerExecutable.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/executable/SchemaCrawlerExecutable.java
@@ -28,6 +28,7 @@ http://www.gnu.org/licenses/
 package schemacrawler.tools.executable;
 
 import static java.util.Objects.requireNonNull;
+import static schemacrawler.tools.utility.SchemaCrawlerUtility.matchSchemaRetrievalOptions;
 import static us.fatehi.utility.Utility.requireNotBlank;
 
 import java.sql.Connection;
@@ -77,7 +78,7 @@ public final class SchemaCrawlerExecutable {
   public void execute() throws Exception {
 
     if (schemaRetrievalOptions == null) {
-      schemaRetrievalOptions = SchemaCrawlerUtility.matchSchemaRetrievalOptions(connection);
+      schemaRetrievalOptions = matchSchemaRetrievalOptions(connection);
     }
 
     // Load the command to see if it is available

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/utility/SchemaCrawlerUtility.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/utility/SchemaCrawlerUtility.java
@@ -44,6 +44,7 @@ import schemacrawler.schema.ResultsColumns;
 import schemacrawler.schemacrawler.DatabaseServerType;
 import schemacrawler.schemacrawler.SchemaCrawlerException;
 import schemacrawler.schemacrawler.SchemaCrawlerOptions;
+import schemacrawler.schemacrawler.SchemaCrawlerRuntimeException;
 import schemacrawler.schemacrawler.SchemaRetrievalOptions;
 import schemacrawler.schemacrawler.SchemaRetrievalOptionsBuilder;
 import schemacrawler.tools.catalogloader.CatalogLoader;
@@ -136,8 +137,7 @@ public final class SchemaCrawlerUtility {
    * @return SchemaRetrievalOptions
    * @throws SchemaCrawlerException On an exception.
    */
-  public static SchemaRetrievalOptions matchSchemaRetrievalOptions(final Connection connection)
-      throws SchemaCrawlerException {
+  public static SchemaRetrievalOptions matchSchemaRetrievalOptions(final Connection connection) {
     final SchemaRetrievalOptionsBuilder schemaRetrievalOptionsBuilder =
         buildSchemaRetrievalOptions(connection);
 
@@ -154,7 +154,7 @@ public final class SchemaCrawlerUtility {
    * @throws SchemaCrawlerException On an exception.
    */
   private static SchemaRetrievalOptionsBuilder buildSchemaRetrievalOptions(
-      final Connection connection) throws SchemaCrawlerException {
+      final Connection connection) {
     checkConnection(connection);
     final DatabaseConnectorRegistry registry =
         DatabaseConnectorRegistry.getDatabaseConnectorRegistry();
@@ -177,11 +177,11 @@ public final class SchemaCrawlerUtility {
     return schemaRetrievalOptionsBuilder;
   }
 
-  private static void checkConnection(final Connection connection) throws SchemaCrawlerException {
+  private static void checkConnection(final Connection connection) {
     try {
       DatabaseUtility.checkConnection(connection);
     } catch (final SQLException e) {
-      throw new SchemaCrawlerException("Bad database connection", e);
+      throw new SchemaCrawlerRuntimeException("Bad database connection", e);
     }
   }
 

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/utility/SchemaCrawlerUtility.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/utility/SchemaCrawlerUtility.java
@@ -201,7 +201,8 @@ public final class SchemaCrawlerUtility {
       final ConnectionInfo connectionInfo = ConnectionInfoBuilder.builder(connection).build();
       LOGGER.log(Level.INFO, connectionInfo.toString());
     } catch (final SQLException e) {
-      LOGGER.log(Level.WARNING, "Could not log connection information", e);
+      LOGGER.log(Level.WARNING, "Could not log connection information");
+      LOGGER.log(Level.FINE, "Could not log connection information", e);
     }
   }
 

--- a/schemacrawler-tools/src/test/java/schemacrawler/test/SchemaCrawlerUtilityTest.java
+++ b/schemacrawler-tools/src/test/java/schemacrawler/test/SchemaCrawlerUtilityTest.java
@@ -46,6 +46,7 @@ import schemacrawler.schema.Catalog;
 import schemacrawler.schema.ResultsColumns;
 import schemacrawler.schema.Schema;
 import schemacrawler.schemacrawler.SchemaCrawlerException;
+import schemacrawler.schemacrawler.SchemaCrawlerRuntimeException;
 import schemacrawler.test.utility.TestDatabaseConnectionParameterResolver;
 import schemacrawler.tools.utility.SchemaCrawlerUtility;
 
@@ -64,7 +65,7 @@ public class SchemaCrawlerUtilityTest {
   public void getCatalogClosedConnection(final Connection connection) throws Exception {
     connection.close();
     assertThrows(
-        SchemaCrawlerException.class,
+        SchemaCrawlerRuntimeException.class,
         () -> SchemaCrawlerUtility.getCatalog(connection, newSchemaCrawlerOptions()));
   }
 

--- a/schemacrawler-tools/src/test/java/schemacrawler/tools/databaseconnector/DatabaseConnectorTest.java
+++ b/schemacrawler-tools/src/test/java/schemacrawler/tools/databaseconnector/DatabaseConnectorTest.java
@@ -39,6 +39,7 @@ import org.junit.jupiter.api.Test;
 import schemacrawler.schemacrawler.SchemaCrawlerException;
 import schemacrawler.schemacrawler.SchemaCrawlerOptions;
 import schemacrawler.schemacrawler.SchemaCrawlerOptionsBuilder;
+import schemacrawler.schemacrawler.SchemaCrawlerRuntimeException;
 import schemacrawler.test.utility.TestDatabaseConnector;
 import schemacrawler.test.utility.WithSystemProperty;
 import schemacrawler.tools.executable.commandline.PluginCommand;
@@ -99,7 +100,7 @@ public class DatabaseConnectorTest {
     final DatabaseConnector databaseConnector = DatabaseConnector.UNKNOWN;
 
     assertThrows(
-        SchemaCrawlerException.class,
+        SchemaCrawlerRuntimeException.class,
         () ->
             databaseConnector.newDatabaseConnectionSource(
                 new DatabaseUrlConnectionOptions("jdbc:mysql://localhost:9001/schemacrawler")));

--- a/schemacrawler-utility/src/main/java/us/fatehi/utility/string/StringFormat.java
+++ b/schemacrawler-utility/src/main/java/us/fatehi/utility/string/StringFormat.java
@@ -29,15 +29,20 @@ package us.fatehi.utility.string;
 
 import java.util.Formatter;
 import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 import us.fatehi.utility.Utility;
 
 public final class StringFormat implements Supplier<String> {
 
+  private static final Logger LOGGER = Logger.getLogger(StringFormat.class.getName());
+
   private final Object[] args;
   private final String format;
 
   public StringFormat(final String format, final Object... args) {
+    // Be tolerant - allow null or blank format strings
     this.format = format;
     this.args = args;
   }
@@ -50,6 +55,11 @@ public final class StringFormat implements Supplier<String> {
 
     try (final Formatter formatter = new Formatter()) {
       return formatter.format(format, args).toString();
+    } catch (final Throwable e) {
+      // NOTE: Do not output arguments, since the toString on argument may throw an exception
+      // obscuring this one
+      LOGGER.log(Level.FINEST, String.format("Error logging message <%s>", format));
+      return "";
     }
   }
 

--- a/schemacrawler-utility/src/test/java/us/fatehi/utility/test/string/StringFormatTest.java
+++ b/schemacrawler-utility/src/test/java/us/fatehi/utility/test/string/StringFormatTest.java
@@ -31,9 +31,6 @@ package us.fatehi.utility.test.string;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.nullValue;
-import static org.junit.jupiter.api.Assertions.assertThrows;
-
-import java.util.IllegalFormatConversionException;
 
 import org.junit.jupiter.api.Test;
 
@@ -43,8 +40,7 @@ public class StringFormatTest {
 
   @Test
   public void badFormat() {
-    assertThrows(
-        IllegalFormatConversionException.class, () -> new StringFormat("%d", "hello").get());
+    assertThat(new StringFormat("%d", "hello").get(), is(""));
   }
 
   @Test

--- a/schemacrawler-website/src/site/markdown/offline.md
+++ b/schemacrawler-website/src/site/markdown/offline.md
@@ -1,32 +1,32 @@
-# SchemaCrawler Offline Snapshot
+# SchemaCrawler Offline Catalog Snapshot
 
 SchemaCrawler allows you to save off your database metadata into an 
-offline snapshot, for future use, using the [`serialize`](serialize.html) command. 
-Later, you can connect to this offline snapshot as you would to a regular database. 
+offline catalog snapshot, for future use, using the [`serialize`](serialize.html) command. 
+Later, you can connect to this offline catalog snapshot as you would to a regular database. 
 This allows you to store historical versions of your schema, or to query your 
 database even if you no longer have access to it. Please make sure that you use
 the same version of SchemaCrawler to serialize snapshots as you use when you 
-load those offline snapshots. Also make sure that the serialization was done using
+load those offline catalog snapshots. Also make sure that the serialization was done using
 the binary Java serialization format, and not the YAML or JSON formats.
 
-You can use any of the rich SchemaCrawler functionality with offline 
+You can use any of the rich SchemaCrawler functionality with offline catalog 
 snapshots, including grep, diagramming, scripting, templating and 
 producing output in a variety of formats. 
 
-## How to Create an Offline Snapshot
+## How to Create an Offline Catalog Snapshot
 
 Use the [`serialize`](serialize.html) command with `--output-format=ser` to 
 serialize the database schema metadata into a binary Java serialization
 format file. It is best to create the snapshot with the `--info-level=maximum`
 and no [limit, filter or grep options](schemacrawler-shell.html).
 
-*Important:* The offline snapshot should should have been created with 
+*Important:* The offline catalog snapshot should should have been created with 
 the same version of SchemaCrawler that you will use to load it.
 
 
-## How to Use an Offline Snapshot
+## How to Use an Offline Catalog Snapshot
 
-In order to connect to an offline snapshot, use the `offline` database
+In order to connect to an offline catalog snapshot, use the `offline` database
 server type, and then use any SchemaCrawler command that you would like
 to use. You can apply [limit, filter or grep options](schemacrawler-shell.html)
 too.

--- a/schemacrawler-website/src/site/markdown/serialize.md
+++ b/schemacrawler-website/src/site/markdown/serialize.md
@@ -43,5 +43,5 @@ to version.
 When you use Java serialization (`--output-format=ser`) to serialize database
 metadata to a file, you can load a serialized metadata using the [`offline`](offline.html)
 server type. Please make sure that you use the same version of SchemaCrawler to 
-create serialized snapshots and to load those offline snapshots.
+create serialized snapshots and to load those offline catalog snapshots.
  

